R/utilities-figures.R

```diff
@@ -223,3 +223,129 @@ createEsqlabsExportConfiguration <- function(projectConfiguration) {
 
   return(exportConfiguration)
 }
+
+#' Generate plots as defined in excel file `projectConfiguration$plotsFile`
+#'
+#' @param simulatedScenarios A list of simulated scenarios as returned by `runScenarios()`
+#' @param observedData A list of `DataSet` objects
+#' @param projectConfiguration Object of class `ProjectConfiguration`
+#' that contains information about the output paths and the excel file
+#' where plots are defined.
+#'
+#'
+#' @return A list of `ggplot` objects
+#'
+#' @export
+plotsFromExcel <- function(simulatedScenarios, observedData, projectConfiguration) {
+  validateIsOfType(observedData, DataSet)
+  validateIsOfType(projectConfiguration, ProjectConfiguration)
+  dfDataCombined <- readExcel(projectConfiguration$plotsFile, sheet = "DataCombined")
+
+  # warnings for invalid data in plot definitions from excel
+  # scenario not present in simulatedScenarios
+  missingScenarios <- setdiff(setdiff(dfDataCombined$scenario, names(simulatedScenarios)), NA)
+  if (length(missingScenarios) != 0) {
+    warning(messages$warningInvalidScenarioName(missingScenarios, projectConfiguration$plotsFile))
+    dfDataCombined <- dfDataCombined[!(dfDataCombined$scenario %in% missingScenarios),]
+  }
+  # data set name not present in observedData
+  missingDataSets <- setdiff(setdiff(dfDataCombined$dataSet, names(observedData)), NA)
+  if (length(missingDataSets) != 0) {
+    warning(messages$warningInvalidDataSetName(missingDataSets, projectConfiguration$plotsFile))
+    dfDataCombined <- dfDataCombined[!(dfDataCombined$dataSet %in% missingDataSets),]
+  }
+
+  # create named list of DataCombined objects
+  dataCombinedNames <- unique(dfDataCombined$DataCombinedName)
+  dataCombinedList <- list()
+  for (name in dataCombinedNames) {
+    dataCombinedList[[name]] <- DataCombined$new()
+    # add data to DataCombined object
+    # add simulated data
+    simulated <- dplyr::filter(dfDataCombined, DataCombinedName == name, dataType == "simulated")
+    if (nrow(simulated) > 0) {
+      for (j in 1:nrow(simulated)) {
+        dataCombinedList[[name]]$addSimulationResults(
+          simulationResults = simulatedScenarios[[simulated[j,]$scenario]]$results,
+          quantitiesOrPaths = simulated[j,]$path,
+          groups = simulated[j,]$group,
+          names = simulated[j,]$label)
+      }
+    }
+
+    # add observed data
+    observed <- dplyr::filter(dfDataCombined, DataCombinedName == name, dataType == "observed")
+    if (nrow(observed) > 0) {
+      dataSets <- observedData[observed$dataSet]
+      dataCombinedList[[name]]$addDataSets(dataSets, groups = observed$group)
```

Comment:
We will need tests for:

- no groups specified for observed data
- Same group for all obs data
- Different groups
- Some obs data with groups, some without.